Reorder UI, add panner, add hide button

alshakero opened the pull request:
I added the SVG for the hiding button, if you want to use the same button as a show/hide toggle, you can just hide/show
`<line stroke-width="50" x1="300" y1="0" x2="60" y2="350" />
` 
element inside the SVG:

```
<svg x="0px" y="0px" width="16px" height="16px" viewBox="0 -20 300 400">
    <path d="M359 204c4 -2 6 -6 6 -11c0 -4 -2 -8 -4 -10v0v0l-1 -1l-79 -79v0c-24 -27 -60 -44 -99 -44c-35 0 -66 14 -90 36v-1l-87 87c-3 3 -5 7 -5 11c0 3 1 7 3 9v0l80 79c24 27 60 45 99 45c35 0 67 -14 91 -36v1zM182 102c50 0 90 40 90 90s-40 90 -90 90s-90 -40 -90 -90 s40 -90 90 -90zM139 192c0 29 14 43 43 43s44 -14 44 -43s-15 -44 -44 -44s-43 15 -43 44z"></path>
   --> <line stroke-width="50" x1="300" y1="0" x2="60" y2="350" />
</svg>
```

tomalec commented on starcounter-layout-html-editor.html:
```diff
@@ -219,6 +240,45 @@
+                    /* reset on double click */
+                    panner.addEventListener('dblclick', function(ev) {
+                        leftPanel.style.flexGrow = 0.5;
+                        rightPanel.style.flexGrow = 0.5;   
```
Why did you add event listener to the entire window?

alshakero replied:
Because when you move the mouse to drag, you can sometimes miss the _panner_ div, and it will not be the event target, and your callback wouldn't be called. You need the dragging to work even if your pointer leaves the div, tor example when you go below/above it.

tomalec replied:
`drag` events should refer correct target even when moved ouside

tomalec commented on juicy-simple-visual-html-editor.html:
```diff
@@ -13,16 +13,18 @@
-                min-height: 1em;
-                display: flex;
-                flex-direction: column;
+           juicy-dom-tree-view{
```
Why have you removed `:host` styles? At least, we need it to be non inline element by default. min height also makes sense to me

alshakero replied:
Because when you set `display` to `flex` the tools will not be fixed and this will happen 
![image](https://cloud.githubusercontent.com/assets/17054134/21658684/d6884b8c-d2c7-11e6-8122-1883ca86f1a1.png)
and if you scroll even more you won't be able to use the tools for lower nodes. And I couldn't see the need of minimum height, but wouldn't be of harm of course. 

That's because when display is flex you can't say overflow to hidden (http://stackoverflow.com/questions/21515042/scrolling-a-flexbox-with-overflowing-content)

alshakero replied:
I'd be happy if you suggest something.

tomalec commented:
Please attach listeners Polymer way https://meowni.ca/posts/polymer-cheatsheet/#listeners

tomalec commented on juicy-simple-visual-html-editor.html:
```diff
@@ -14,15 +14,20 @@
-                min-height: 1em;
-                display: flex;
-                flex-direction: column;
+                min-height: 2em;
```
I still don't get the point why flex is no good in here
![image](https://cloud.githubusercontent.com/assets/17435/21661000/4382cbcc-d2d2-11e6-8dee-6ae5ca9069cf.png)
We do not need `overflow:hidden` here. I'd say `display:inline` is even worse.

alshakero replied:
Can you scroll down the tree and share a screenshot?

tomalec replied:
![image](https://cloud.githubusercontent.com/assets/17435/21663316/b45807e2-d2df-11e6-8f9c-387ae229a056.png)

tomalec commented on starcounter-layout-html-editor.html:
```diff
@@ -220,6 +271,9 @@
                     });
                     this.refreshSelectors();
+
+                    window.addEventListener('mouseup', this.windowMouseUp.bind(this));
```
If you are attaching listeners manually you need to take care of detaching as well

tomalec commented on starcounter-layout-html-editor.html:
```diff
@@ -98,7 +119,37 @@
+                    this.isMouseDown = true;
+                },
+                windowMouseMove: function(ev) {
+                    if(this.isMouseDown)
```
Have you tried drag events?

alshakero replied:
Trying now. They were always pain in the neck, I'm giving them another shot if you think that's better.

tomalec commented:
In general I'd prefer to use `em` rather than `px` as then it could behave nicer if you would style entire context

tomalec commented:
I'd also add some white-space between separator and the editors, in general I don't find it intuitive that it's draggable, maybe we could get some inspiration from  https://material.io/guidelines/layout/split-screen.html

Could you make those tile collapsible by separate button/arrow like https://material.io/guidelines/components/expansion-panels.html
Personally I would find it nicer to collapse it by single click, rather than dragging it to the very end.
Also once collapsed, we could hide the divider, and just leave small icon & vertical text to expand

alshakero commented:
@tomalec you're right, it's not intuitive, I'll look at the reference you mentioned. 

As for collapsing the panels. I actually implemented it at first. But I figured it's a bit more complicated than its worth. Think about how many possibilities there needs to be.

1. Default state, both visible, the separator is in the middle, show both collapse buttons in _hide_ mode. Easy.
2. One is collapsed, the other should hide its collapse button, to avoid collapsing both. 
3. Both are visible, but not 50/50%. Say 30/70%. Then one collapse is clicked. We should make the visible one 100% and preserve the 30/70% when the user shows the hidden one again. 

And at least 3 other scenarios for a functionality that is already there by simply dragging. 

If you really like a collapse option, I think a separate radio group would be perfect. Something like:

- [x]  Show both
- [ ]  Show right
- [ ]  Show left




Way easier to implement, what do you think?

tomalec commented:
Sorry, I don't get what you mean by all 1-3, maybe HO call could help.

I still don't get why it's complicated, but for sure I find yet another radio button very unintuitive.

warpech commented:
What is the status of this? Should it wait until @tomalec is back from vacation?

alshakero commented:
@warpech oh, it was a rush and both @tomalec and I implemented the collapsing feature by accident. However, I also implemented the drag resize with it, but it was on Wednesday and Tomek preferred to use his code rather than reviewing mine for the tight schedule and his vacation approaching.

warpech commented:
So I propose that we try to finish PR #13 and release a version of CompositionEditor with it to meet the schedules.

Then you can decide about PR #12 when @tomalec is back from vacation. 

WDYT?

alshakero commented:
Closing as implimented by @tomalec .
